File: ndk-mobile/expo-ndk-mobile-nip-55/src/ExpoNDKMobileNip55Module.ts
import { /* NativeModule, */ requireNativeModule } from "expo";

// import { ExpoNDKMobileNip55ModuleEvents } from './ExpoNDKMobileNip55.types';

// declare class ExpoNDKMobileNip55Module extends NativeModule<ExpoNDKMobileNip55ModuleEvents> {
//   PI: number;
//   hello(): string;
//   setValueAsync(value: string): Promise<void>;
// }

declare class ExpoNDKMobileNip55Module {
    getTheme(): string;
    getPublicKey(): string;
}

// This call loads the native module object from the JSI.
export default requireNativeModule<ExpoNDKMobileNip55Module>("ExpoNDKMobileNip55");

File: ndk-mobile/expo-ndk-mobile-nip-55/android/src/main/java/expo/modules/ndk-mobile-nip-55/ExpoNDKMobileNip55Module.kt
package expo.modules.ndk-mobile-nip-55

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import biz.nostr.nip55.Signer;

class ExpoNDKMobileNip55Module : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoNDKMobileNip55')` in JavaScript.
    Name("ExpoNDKMobileNip55")

    // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
    Constants(
      "PI" to Math.PI
    )

    // Defines event names that the module can send to JavaScript.
    Events("onChange")

    // Defines a JavaScript synchronous function that runs the native code on the JavaScript thread.
    Function("hello") {
      "Hello world! 👋"
    }

    Function("getTheme") {
      // return@Function "system"
      "Theme! 👋"
    }

    Function("getPublicKey") {
      "Public key"
      // Signer signer = new Signer(context);
      //
      // try {
      //     String publicKey = signer.getPublicKey();
      //     // Use the public key as needed
      // } catch (Exception e) {
      //     // Handle exception
      // }
    }

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("setValueAsync") { value: String ->
      // Send an event to JavaScript.
      sendEvent("onChange", mapOf(
        "value" to value
      ))
    }
  }
}
